Display relative peak intensities in InstrumentView

MichaelWedel opened the pull request:
This addresses #13907.

**Testing instructions**
1. Use the following script to create a workspace and a list of predicted peaks:
```
lattice_parameters = {'a': 5.431, 'b': 5.431, 'c': 5.431 }

CreateSimulationWorkspace(Instrument='WISH', BinParams='0,10000,20000',
                                        OutputWorkspace='test', UnitX='TOF')
                                        
SetUB(Workspace='test', u='12,1,1', v='0,4,9', **lattice_parameters)

cs = CrystalStructure('{0} {1} {2}'.format(*lattice_parameters.values()), 'F d -3 m', 'Si 0 0 0 1.0 0.01')
f = mtd['test']
f.sample().setCrystalStructure(cs)

PredictPeaks(InputWorkspace='test', CalculateStructureFactors=True,
                    MinDSpacing=0.5, MaxDSpacing=100, OutputWorkspace='predicted',
                    ReflectionCondition='All-face centred')
```
2. Open the instrument view on the `test` workspace.
3. Drag the peaks workspace onto the instrument display. All spots should have the same size.
4. Toggle the "Indicate relative intensity" option in the "Peak options" menu of the "Render" tab and notice that the spot sizes are now different. Make sure that turning it on and off works properly. When you change the `CalculateStructureFactor` option to `False`, all intensities will be zero and there should be no difference between the two options.
5. Make sure that the state of this option is preserved across Mantid restarts.

SimonHeybrock commented:
I don't know if this is an intended way to use this but try the following:

- run script, drag into instrument view, set indicate to true (as you describe it)
- modify script to set `CalculateStructureFactor` to `false`, rerun, drag into instrument view

Now I have two sets of peaks, the ones with relative intensity (red) and the new ones, all same size (green). Good.

Now toggle the indicate rel. intensity flag in the peaks options menu (e.g., off and on again). Now both sets of peaks are red, i.e., the color information is gone.

SimonHeybrock commented on MantidPlot/src/Mantid/InstrumentWidget/PeakOverlay.cpp:
```diff
@@ -242,26 +287,25 @@ void PeakOverlay::draw(QPainter& painter) const
   }
-  //std::cerr << m_labels.size() << " labels " << color.red() << ' ' << color.green() << ' ' << color.blue() << "\n";
+  // std::cerr << m_labels.size() << " labels " << color.red() << ' ' <<
+  // color.green() << ' ' << color.blue() << "\n";
```
Remove?

SimonHeybrock commented on MantidPlot/src/Mantid/InstrumentWidget/PeakOverlay.cpp:
```diff
@@ -242,26 +287,25 @@ void PeakOverlay::draw(QPainter& painter) const
+  for (int i = 0; i < m_labels.size(); ++i) {
+    PeakHKL &hkl = m_labels[i];
+    hkl.draw(painter, m_precision);
+    // hkl.print();
```
Remove?

MichaelWedel commented:
I think I know where the color information is lost, I'm fixing it right now. Thanks for spotting it. At the same time I'll remove the commented code. That may have been there for some time, I accidentally clang-formatted the files so it looks like I changed more than I actually did :)

SimonHeybrock commented on MantidPlot/src/Mantid/InstrumentWidget/PeakOverlay.h:
```diff
@@ -28,50 +28,110 @@ class UnwrappedSurface;
+  AbstractIntensityScale(
+      const boost::shared_ptr<Mantid::API::IPeaksWorkspace> &pws) {
+    setPeaksWorkspace(pws);
+  }
```
Shouldn't there be a virtual destructor? You have pointers to child classes, I think.

PascalManuel replied:
What is done exactly about color for peak intensities ?

Colors are also used for different UB in the instrument view so it might lead to confusion.



________________________________
From: Simon Heybrock [notifications@github.com]
Sent: 13 January 2016 8:52 AM
To: mantidproject/mantid
Subject: Re: [mantid] Display relative peak intensities in InstrumentView (#14935)


In MantidPlot/src/Mantid/InstrumentWidget/PeakOverlay.h<https://github.com/mantidproject/mantid/pull/14935#discussion_r49563670>:

>    QList<int> rows; ///< row indices of the peaks in their PeaksWorkspace
>    bool showRows;
>  };
>
> +/// Helper class for scaling peak markers to intensities.
> +class AbstractIntensityScale {
> +public:
> +  AbstractIntensityScale(
> +      const boost::shared_ptr<Mantid::API::IPeaksWorkspace> &pws) {
> +    setPeaksWorkspace(pws);
> +  }


Shouldn't there be a virtual destructor? You have pointers to child classes, I think.

�
Reply to this email directly or view it on GitHub<https://github.com/mantidproject/mantid/pull/14935/files#r49563670>.

MichaelWedel replied:
@PascalManuel Color is not touched for intensity, there was just a problem that color information was lost when switching between modes.

SimonHeybrock commented:
Regarding "Make sure that the state of this option is preserved across Mantid restarts.": The state is preserved (i.e., I get intensity indication after restart), but not the state of the menu (it is not set, if I set nothing happens, if I unset again it unsets).

SimonHeybrock commented on MantidPlot/src/Mantid/InstrumentWidget/PeakOverlay.h:
```diff
@@ -28,50 +28,110 @@ class UnwrappedSurface;
+  /// Returns the base style unmodified.
+  PeakMarker2D::Style
+  getScaledMarker(double intensity,
+                  const PeakMarker2D::Style &baseStyle) const {
```
`override` should be supported by all our compilers now? Then I'd prefer writing `const override {`.

SimonHeybrock commented on MantidPlot/src/Mantid/InstrumentWidget/PeakOverlay.h:
```diff
@@ -28,50 +28,110 @@ class UnwrappedSurface;
+
+protected:
+  PeakMarker2D::Style
+  getScaledMarker(double intensity, const PeakMarker2D::Style &baseStyle) const;
```
`override`

SimonHeybrock commented on MantidPlot/src/Mantid/InstrumentWidget/ProjectionSurface.h:
```diff
@@ -285,6 +287,7 @@ protected slots:
   mutable int m_peakLabelPrecision;
   mutable bool m_showPeakRows;        ///< flag to show peak row index
   mutable bool m_showPeakLabels;     ///< flag to show peak hkl labels
+  mutable bool m_showPeakRelativeIntensity;     ///< flag to show peak hkl labels
```
Why does this have to be `mutable`?

MichaelWedel replied:
I simply assumed there is a reason the others are declared `mutable`.

SimonHeybrock commented:
There might be, but I could not see it. I guess the compiler should complain if he really needs it?

MichaelWedel commented:
I tried compiling without `mutable`, it works, I could not see any change in behavior either. Still I'm somehow reluctant to just change this without knowing more about why it was put there. My suggestion is to leave it like that for now and try to clarify why it was written like this in the first place at a later point. There will be opportunities for this anyway, with the widget moving to MantidQt and so on.

MichaelWedel commented on MantidPlot/src/Mantid/InstrumentWidget/InstrumentWindowRenderTab.cpp:
```diff
@@ -567,6 +567,12 @@ QMenu* InstrumentWindowRenderTab::createPeaksMenu()
 
+  QAction *indicateRelativeIntensity = new QAction("Indicate relative intensity", this);
+  indicateRelativeIntensity->setCheckable(true);
+  indicateRelativeIntensity->setChecked(settings.value("IndicateRelativeIntensities", false).toBool());
```
It seems this is not enough to restore the menu state. I'll try to find out if the menu is modified elsewhere.

MichaelWedel replied:
Turns out this was my fault with inconsistent naming of options.

SimonHeybrock commented:
Well, correct me if I am wrong, but the only reason for mutable is if it needs to be changed by a `const` method. And if it is, there should by a lock or mutex protecting such a change (since in C++11 the compiler assumes that `const` methods are thread safe). So I see no harm in not putting it, having it is potentially dangerous.

MichaelWedel commented:
@SimonHeybrock: The behavior-problems you discovered should be fixed now. I removed `mutable` from the member I added, but changing the other ones will have to happen at another point.

@PascalManuel: Could you check if this is the way you imagined this to be implemented, at least as a first iteration for the functionality?

PascalManuel commented:
Will do.



Dr Pascal Manuel
ISIS Pulsed Neutron and Muon Facility
STFC Rutherford Appleton Laboratory
Chilton
Oxfordshire OX11 0QX
United Kingdom
Office: + 44 (0)1235 446290
Beamline : +44 (0)1235 567080
Mobile: +44(0) 7833 965011

________________________________
From: Michael Wedel [notifications@github.com]
Sent: 13 January 2016 10:14 AM
To: mantidproject/mantid
Cc: Manuel, Pascal (STFC,RAL,ISIS)
Subject: Re: [mantid] Display relative peak intensities in InstrumentView (#14935)


@SimonHeybrock<https://github.com/SimonHeybrock>: The behavior-problems you discovered should be fixed now. I removed mutable from the member I added, but changing the other ones will have to happen at another point.

@PascalManuel<https://github.com/PascalManuel>: Could you check if this is the way you imagined this to be implemented, at least as a first iteration for the functionality?

�
Reply to this email directly or view it on GitHub<https://github.com/mantidproject/mantid/pull/14935#issuecomment-171241820>.

MichaelWedel commented:
I will update the release notes once this is merged.

SimonHeybrock commented:
Merging this, after Pascal gave his go.
